crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4185,6 +4301,89 @@ mod tests {
         Ok(())
     }
 
+    #[test]
```

Comment:
Thanks for adding thorough tests!

A few tests I'd like to add:

* Some tests of the diagnostic we get for unsupported comparisons inside a tuple, with TODO comment to improve that diagnostic, as discussed above.
* A test for the case shown above where `__eq__` returns a non-bool. (This will just be a Todo for now, unless we implement Type::Instance rich comparison support for `__eq__`, but it will be useful to have the test in place for future anyway.)

Reply:
- unsupported comparison
https://github.com/astral-sh/ruff/pull/13712/files#diff-eff9f162b9d83ddfe43d3f6b28ed022767bc58d12d7ccb0c95d985e827471362R83-R103
- non-bool 
(https://github.com/astral-sh/ruff/pull/13712/files#diff-eff9f162b9d83ddfe43d3f6b28ed022767bc58d12d7ccb0c95d985e827471362R83-R103)